design_documents/parametrizing_tests_with_fixed_values.md

```diff
@@ -0,0 +1,192 @@
+# Parametrizing Tests With Fixed Values
+
+## Context
+
+Various test runners for other programming languages provide ways of parametrizing test cases with fixed values.
+This way, multiple test cases can be executed for the same test source code.
+
+## Goal
+
+Propose a way of adding support for parametrized tests to Starknet Foundry's `snforge`.
+
+## Proposed Solution
+
+Introduce two new attributes: `parametrize` and `case`.
+
+- `parametrize` - defines a parametrized test.
+- `case` - defines a specific test case
+
+Generate code for each of the test cases so no handling of values injection, etc. is necessary.
+The compiler would then validate if the values for the generated code are actually valid.
+Thanks to that, we could have failures for invalid arguments early in the execution.
+
+### Parametrization With Simple Arguments
+
+```cairo
+#[parametrize]
+#[case(1, 3)]
+#[case(3, 5)]
+#[test]
+fn my_test(a: felt252, b: u32) {
+    // case a = 1, b = 3
+    // case a = 3, b = 5
+    // ...
+}
+```
+
+For test cases like this, we would generate two actual test cases:
+
+```cairo
+#[test]
+fn my_test_case_1() {
+  let a: felt252 = 3;
+  let b: u32 = 1;
+}
+
+#[test]
+fn my_test_case_1() {
+  let a: felt252 = 3;
+  let b: u32 = 5;
+}
+```
+
+The explicit type annotations for each argument are necessary.
+This way, if the user provides invalid value for the type defined, the compiler will throw a relevant error.
+
+### Parametrization With Fixtures
+
+```cairo
+struct MyStruct {
+    a: felt252, 
+    b: Array<felt252>
+}
+
+fn my_fixture() -> MyStruct {
+    MyStruct { a: 1, b: array![2, 3] }
+}
```

Comment:
Missing details on:
- more advanced use-cases and how this solution plays along with them. What are common use cases, how this solution handles them? (imo not very well)
- I have to write a function for every case even if it is fixed, this is a load of boilderplate and kinda kills the purpose, it is easier to just write a normal test.
- How this will be type checked? What are the constraints of fixture functions?

Reply:
Imo generally you underestimate the complexity of this solution, if you have provided more details you would run into more problems. It might just look plausible.

Reply:
> more advanced use-cases and how this solution plays along with them. What are common use cases, how this solution handles them? (imo not very well)

What do you mean by more advanced use cases? 

> I have to write a function for every case even if it is fixed, this is a load of boilderplate and kinda kills the purpose, it is easier to just write a normal test.

If your function takes simple values as arguments then you don't need a fixture. Requiring them for complex types is mostly so it's easier to implement. 

> How this will be type checked? What are the constraints of fixture functions?

I don't think annotations are type checked at all. We'd have to have our own handling (can be discussed in this document).

Reply:
> What do you mean by more advanced use cases?

Real life use-cases, example of a test which could exists/exist in some real codebase and its boilerplate would be reduced

> We'd have to have our own handling (can be discussed in this document).

If you propose something like this you should discuss them, I assume those are not type checked by the compiler.

Reply:
I think such simple fixtures don't provide much benefit and cloud the scope of this task, I think it is another feature that could be considered, but I wouldn't do it there. 
Fixtures start to be beneficial if they are supported on wider range like in pytest imo